### DISCUSSION TITLE
Exact time trigger for Websend needed

### DISCUSSION THREAD

**Discussion by schleckew** [arendst/Tasmota#19220] [Q&A]
I have two devices sending data to the same Thingspeak Channel. 
The time interval between two messages must be minimum 15 seconds otherwise the later websend becomes ignored. 
Device 1 is sending minute exact data by reliable time trigger rule "On Time#Minute.." (very reliable). 
Device 2 should send via Script every minute 2 data with 20 and 40 seconds offset to Device 1 sent data.
For Device 2 I used basically 20 seconds upsecs trigger skipping the Websend around Time#Minute trigger of Device 1:
">S
if upsecs%20=0..."

Unfortunately the upsecs timer is unreliable and once a day the trigger shifts by several seconds causing overlapping Websend of Device 1 and Device 2. Partially the minute based data of Device 1 is then not send at all.

Can anybody help to built reliable trigger for Websend on Device 2 Script?

**Comment by barbudor** [comment]
Do you need to use script for other reasons?
Using rules I would use the same trigger than on device 1 to start 2 Rulestimer, one with 20 sec and one with 40 sec

Alternatively can also have device 1 rule the sequence with those ruletimer and sending either an event command or event the WebSend command through device 2

**Comment by schleckew** [comment]
Thanks for fast reply!
I need script on the ESP8266 bitShake SmartReader to filter some data of household electricity. Unfortunately I have no smart home PC to synchronize with 2nd device (

**Reply by schleckew** [reply]
Wlan ESP8266 smart plug)
So there is no sync possibility between both devices. That's why I propose absolute timer based solution with 20 Seconds based offset.

**Reply by sfromis** [reply]
You could do something like having the time trigger rule every minute use a RuleTimer to schedule another rule 20 seconds (or whatever) second later, and there publish a command for the device with scripting to take action. Not using scripting, I've got no solid suggestion on what works best, but it might be something like publishing a Script command, or having the script have a subroutine acting as as command.

**Reply by pkkrusty** [reply]
Yeah, use the first esp8266 to manage all the timings, and then subscribe to the topic on the second and trigger the web send off of that. 

Or like sfromis said, issue a command directly from the first ESP to the second, that would bypass the MQTT server. 

Scripting has the possibility of reading and writing global variables, that are shared between all ESP devices, but not sure if both devices would need scripting in that case.

**Reply by sfromis** [reply]
> Or like sfromis said, issue a command directly from the first ESP to the second, that would bypass the MQTT server.

I was not suggesting to bypass the MQTT broker, as that would be detrimental to reliability and speed. My usage of the word "publish" was a reference to the MQTT terms. What I was suggesting takes no extra subscribe, as it would be possible to use the existing `Script` command, or use the scripting feature of having a subroutine called as a command.

**Comment by schleckew** [comment]
Thanks for all your feedback. I will try to setup some rules on the 1st device managing 2 variables to be set at 20 sec and 40 sec offset based on Time#Minute Timer trigger. I hope I can use these variables in the script for reliable Websend trigger.
Keep you updated!

**Comment by schleckew** [comment]
Unfortunately I cannot use Rules in the Console as I was used to do on the 2nd device (Wlan ESP8266 smart plug). Does anybody know, where to place the Rules in the Script?

**Reply by sfromis** [reply]
You cannot combine rules and scripting on the same device, these features are mutually exclusive. But you wrote that you are using rules on device 1, and then you could do the timing there, and forward a command to device 2 to do the needed things there, instead of trying to manage the timing on device 2. 

There may very well be a way of doing it with scripting on device 2, but as I'm not using this feature, I cannot make suggestions about how.

**Comment by schleckew** [comment]
Seems I can only use Script with unreliable timer trigger in device 1 and Rules via Console with reliable Time#Minute trigger in device 2.

**Comment by schleckew** [comment]
Sorry for mixing the device numbers 1 and 2! Obviously the Wlan Socket with reliable rules(originally described as device 1) cannot use Scripts and the bitShake esp8266 need to run with Script exclusively.Now I would send from the Script a command to device 1 to trigger the minute based websend. At the moment I don't know how to send the command to the device in the same network (192.168.1.x). Furthermore how to build the rule trigger based on the sent command. Mit freundlichen GrüßenEdgar Weiter

**Reply by sfromis** [reply]
Well, you already know how you can send a command to another device, since you mention WebSend. Or if on MQTT, you can do a publish. You can check the rules cookbook for how it works with a RuleTimer, which you can do for delayed action following up on the Time#Minute. I cannot go into detail about how to control things, as I do not know your setup, and becomes even less clear what you mean with device 1 and 2.

**Comment by pkkrusty** [comment]
According to scripting docs, >S section is least precise, upsecs next best, and using t:timer1=20 is most precise.  Maybe try using script timers.  I have used the ms timers and that seemed pretty accurate. Section execution was within 2ms and could be calibrated.  But really, best bet would be to trigger device 2 from Device 1.  so

```
rule1
on time#minute do backlog websend whatever; ruletimer1 20 endon
on rules#timer=1 do backlog publish device2/cmnd/power1 2;  ruletimer2 20 endon
on rules#timer=2 do publish device2/cmnd/power1 2 endon
```

define device2 with a dummy relay so you can use the trigger in section >E
```
>D
toggler=0
>E
toggler=Power1
>F
if chg[toggler] {
=>websend whatever
}
```

**Reply by schleckew** [reply]
I cannot use Tasmota MQTT with Thingspeak. Therefor I am using the "Websend" command to send data to ThingSpeak. Instead of performing Websend concurrently from 2 devices to the same Thingspeak channel (min. 15seconds delay between 2 send!) I would like to do commonly from device 2 with script:
**Device 1 (SolarSocket 192.168.1.213) :**
_Rule1
ON Energy#Power DO Websend [192.168.1.213] power_solar=%value% ENDON

Rule2
ON Energy#Total DO Websend [192.168.1.213] total_solar=%value% ENDON_

**Device 2 (Household meter ESP8266 192.168.1.80) :**
Script:
 '>D

power_solar=0
total_solar=0

'>S
 
print power_solar : %power_solar%
print total_solar : %total_solar%

if upsecs%20==0 
...
WebSend [api.thingspeak.com] /update.json?api_key=xxx&field1=%haus_aktuell%&field2=%sml[5]%&field3=%sml[1]%&field4=%sml[4]%&field5=%sml[2]**&field6=%power_solar%&field7=%total_solar%**&field8=%einspeisung_akluell%_**

The Websend from Device 1 seems to work :
_18:30:24.105 RUL: ENERGY#POWER performs "Websend [192.168.1.213] power_solar=67"
18:30:24.191 RSL: RESULT = {"WebSend":"Done"}_

But Device 2 cannot deal with the above Websend command:
_18:32:15.596 RSL: RESULT = {"Command":"Unknown"}
18:32:16.173 RSL: RESULT = {"Command":"Unknown"}
18:32:16.372 power_solar : 0.00
18:32:16.374 total_solar : 0.00
18:32:16.775 RSL: RESULT = {"Command":"Unknown"}
18:32:17.373 power_solar : 0.00
18:32:17.375 total_solar : 0.00
18:32:17.427 RSL: RESULT = {"Command":"Unknown"_

I really looking now for solution according above concept. In this way I can send all 8 data from 2 devices every 20 seconds using upsecs%20 in the script of Device 2.

Thank you very much for your support

**Reply by sfromis** [reply]
Better run with `Weblog 3` activated on device 2 to see details of what is seen as an unknown command. If the 192.168.1.213 you do a websend to is "device 2", that is certainly an invalid command. What are you trying to do?

What I was suggesting earlier was that you could possibly use the [Script](https://tasmota.github.io/docs/Scripting-Language/#console-commands) command, but that's not what you're doing.

**Reply by pkkrusty** [reply]
Yeah you can't refer to script variables directly. As sfromis said you need to websend "script >total_solar=%value%"

**Reply by pkkrusty** [reply]
Are you saying that you have no local MQTT broker? So instead of publishing to MQTT server with device 1 using `publish`, just use `websend [192.168.1.213] script >total_solar_device1=%value%` and with device 2:
```
>D
total_solar_device1=0

>F
if chg[total_solar_device1] {
=>websend whatever
}

**Comment by schleckew** [comment]
Sorry, wrong adresses in the description of Device 1 and Device 2. Below the corrected:
Device 1 (SolarSocket 192.168.1.**63**) :
_Rule1
ON Energy#Power DO Websend [192.168.1.213] power_solar=%value% ENDON

Rule2
ON Energy#Total DO Websend [192.168.1.213] total_solar=%value% ENDON_

Device 2 (Household meter ESP8266 192.168.1.**213**) :
Script:
'>D

power_solar=0
total_solar=0

'>S

print power_solar : %power_solar%
print total_solar : %total_solar%

if upsecs%20==0
...
WebSend [api.thingspeak.com] /update.json?api_key=xxx&field1=%haus_aktuell%&field2=%sml[5]%&field3=%sml[1]%&field4=%sml[4]%&field5=%sml[2]&field6=%power_solar%&field7=%total_solar%&field8=%einspeisung_akluell%_**

As already mentioned the Websend from Device 1 seems to work  (just Rule1 enabled):
18:30:24.105 RUL: ENERGY#POWER performs "Websend [192.168.1.213] power_solar=67"
18:30:24.191 RSL: RESULT = {"WebSend":"Done"}

After the Weblog 3 activation of Device 2 I could confirm the wrong command from the websend of Device 1(192.168.1.63):
_20:16:32.214 HTP: Command
20:16:32.219 SRC: WebCommand from 192.168.1.63
20:16:32.222 CMD: Grp 0, Cmd 'POWER_SOLAR', Idx 1, Len 2, Pld -99, Data '=8'
20:16:32.228 RSL: RESULT = {"Command":"Unknown"}
20:16:32.418 power_solar : 0.00
20:16:32.420 total_solar : 0.00_

It seems that the variable "power_solar" in device 2 is not properly detected for transfer of the value from Device 1(power_solar=67).

Any idea how to get the variables from Device 1 and Device 2 matching for transfer?

**Reply by sfromis** [reply]
Tasmota has no command named "power_solar". You cannot set a variable that way. Did you read what I wrote about using the command "Script"?

**Comment by schleckew** [comment]
I read it, but I cannot understand. 
I am using Script on device 2, so for me there is no need to use Script command in the Console?! 
Or do you want me to use Script command on device 1, where I use Rules via Console ?

**Reply by sfromis** [reply]
The Script command is only workable on the device running scripting instead of rules. And the link I posted has a specific example of how to use the Script command for setting a variable. Use that command with WebSend.

When I write about the "Script" command, I mean a specific **command** with the name "Script", it is not a general reference to the scripting feature/language. And see the link I already posted.

**Comment by schleckew** [comment]
I will try this.

**Comment by schleckew** [comment]
Yes I have no MQTT server. I will try this next week. My house roof collapsed and I have other problems at the moment... Mit freundlichen GrüßenEdgar Weiter

**Reply by schleckew** [reply]
Now I tested according your recommendation:

Device 1 (SolarSocket 192.168.1.63) :
_Rule1
ON Energy#Power DO Websend [192.168.1.213] script >power_solar=%value%

Device 2 (Household meter ESP8266 192.168.1.213) :
Script:
'>D
power_solar=0

>F
if chg[power_solar] {
print power_solar : %power_solar%
}

===========
As I did not get any data on device 2 I tested in the Console of Device 1 the Script command :
Device 1 (SolarSocket 192.168.1.63)  Console:
"script >power_solar=15"

10:08:41.804 CMD: script >power_solar=15
10:08:41.807 SRC: WebConsole from 192.168.1.191
10:08:41.810 CMD: Grp 0, Cmd 'SCRIPT', Idx 1, Len 15, Pld -99, Data '>power_solar=15'
10:08:41.815 RSL: RESULT = {"Command":"Unknown"}

Obviously the Script as console command is not available in Device 1.

**Reply by sfromis** [reply]
You need to execute the Script command on "Device 2" where you have scripting, it makes no sense on "Device 1", outside of sending it to another device using WebSend or whatever.

To see whether the command arrives on "Device 2", you need to watch the console on that device.

**Reply by schleckew** [reply]
I preferred concrete coding sample like from pkkrusty the 
"websend [192.168.1.213] script >total_solar_device1=%value% "
which was obviously meant to be done on Device 1

As Script command works only on Device 2 I am still not sure how to deal with the script command:

**Device 1** (SolarSocket 192.168.1.63) :
Rule1 ON Energy#Power DO Websend [192.168.1.213] power_solar=%value%

_Console:_
14:28:13.677 RUL: ENERGY#POWER performs "Websend [192.168.1.213] power_solar=76"
..

**Device 2** (Household meter ESP8266 192.168.1.213) :
Script:
'>D
power_solar=0

'>F
 if chg[power_solar] {
 print power_solar : %power_solar%
 }
...

Console Script (Device 2):
script >power_solar=%value%

_Console:_
..
14:24:50.722 CMD: script >power_solar=%value%
14:24:50.724 SRC: WebConsole from 192.168.1.191
14:24:50.727 CMD: Grp 0, Cmd 'SCRIPT', Idx 1, Len 20, Pld -99, Data '>power_solar=%value%'
14:24:50.732 var not found: %value%
14:24:50.734 RSL: RESULT = {"Script":">power_solar=%value%"}
14:24:51.280 HTP: Command
14:24:51.285 SRC: WebCommand from 192.168.1.63
14:24:51.288 CMD: Grp 0, Cmd 'POWER_SOLAR', Idx 1, Len 3, Pld -99, Data '=81'
14:24:51.293 RSL: RESULT = {"Command":"Unknown"}

The unknown command comes from Device 1, but if I use the
modified Rule with Websend [192.168.1.213] **script >power_solar=%**value% 
nothing is send from Device 1 to Device 2

If I change the script command on Device 2 to
script >power_solar=20
the print command from the SCRIPT above is not triggered:
14:35:30.402 CMD: script >power_solar=15
14:35:30.404 SRC: WebConsole from 192.168.1.191
14:35:30.407 CMD: Grp 0, Cmd 'SCRIPT', Idx 1, Len 15, Pld -99, Data '>power_solar=15'
14:35:30.412 RSL: RESULT = {"Script":">power_solar=15"}

I would appreciate any corrections in my code above...
